src/buffer.rs

```diff
@@ -239,3 +313,101 @@ pub enum IQMatrix {
     /// Abstraction over `VAIQMatrixBufferHEVC`
     HEVC(hevc::IQMatrixBufferHEVC),
 }
+
+pub enum EncSequenceParameter {
+    H264(h264::EncSequenceParameterBufferH264),
+}
+
+pub enum EncPictureParameter {
+    H264(h264::EncPictureParameterBufferH264),
+}
+
+pub enum EncSliceParameter {
+    H264(h264::EncSliceParameterBufferH264),
+}
+
+pub enum EncMacroblockParameterBuffer {
+    H264(h264::EncMacroblockParameterBufferH264),
+}
+
+pub enum EncCodedBuffer {
+    UndefinedStorage(usize),
+    LocalStorage(Vec<u8>), // This one does not work :C
+}
+
+pub struct CodedBufferSegment<'a> {
+    // pub size: u32,
+    pub bit_offset: u32,
+    pub status: u32,
+    //pub buf: *mut ::std::os::raw::c_void,
+
+    pub buf: &'a [u8],
+}
+
+/// Helper to access segments of mapped coded buffer
+pub struct CodedBuffer<'a> {
+    buffer: Option<Buffer>,
+    segments: Vec<CodedBufferSegment<'a>>,
+}
+
+impl <'a> CodedBuffer<'a> {
+    fn new(buffer: Buffer) -> Result<Self, VaError> {
+        let mut addr = std::ptr::null_mut();
+        let mut segments = Vec::new();
+
+        unsafe {
+            va_check(
+                bindings::vaMapBuffer(buffer.context.display().handle(), buffer.id, &mut addr)
+            )?;
+
+            while addr != std::ptr::null_mut() {
+                let segment =  addr as *mut bindings::VACodedBufferSegment;
+
+                let bit_offset = (*segment).bit_offset;
+                let size = (*segment).size;
+                let buf = (*segment).buf;
+                let status = (*segment).status;
+
+                let buf = std::slice::from_raw_parts(buf as *mut u8, size as usize);
+
+                segments.push(CodedBufferSegment { bit_offset, status, buf });
+
+                addr = (*segment).next;
+            }
+        }
+
+        Ok(Self {
+            buffer: Some(buffer),
+            segments
+        })
+    }
+
+    pub fn segments(&self) -> &Vec<CodedBufferSegment<'a>> {
+        &self.segments
+    }
+
+    pub fn take_buffer(mut self) -> Result<Buffer, VaError> {
```

Comment:
Why?

Reply:
@bgrzesik I do not know any projects that pool the coded side (i.e. `VABuffer`s). All the projects I work with (GStreamer, ffmpeg, Chromium) pool the raw side (i.e. `VASurface`s), because that is what is costly in terms of allocation. Inter-frames in particular can be very light.

Let's drop this idea. If you want an example on `VASurface`pooling in Rust see https://github.com/chromeos/cros-codecs/blob/000e9fa155581da06fe9a382c524518b7f46c169/src/backend/vaapi.rs#L309

Reply:
Sure, removed. Actually that `Option<>` was inspired from that, namely `PooledSurface<>` :smiley: